src/Analyser/MutatingScope.php

```diff
@@ -4379,6 +4379,14 @@ private function shouldInvalidateExpression(string $exprStringToInvalidate, Expr
 		$nodeFinder = new NodeFinder();
 		$expressionToInvalidateClass = get_class($exprToInvalidate);
 		$found = $nodeFinder->findFirst([$expr], function (Node $node) use ($expressionToInvalidateClass, $exprStringToInvalidate): bool {
+			if (
+				$exprStringToInvalidate === '$this'
+				&& $node instanceof Name
+				&& in_array($node->toLowerString(), ['self', 'static', 'parent'], true)
```

Comment:
I just realized the `self` here should be improved. We should use `Scope::resolveName` or `resolveTypeByName` and check that the class name matches the current scope class name. Because `self` is just an alias to the current class name, so whether we're using `self::` or `Foo::` shouldn't matter.